src/test/tcp/with_q.sh

```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+# Catch failures
+set -euo pipefail
+
+# Create Q, and export into environment of child processes
+export QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'`
```

Comment:
You left off the bits of my suggested script that a) clean up the queue. It makes sense to clean it up in the script instead of inside the test to keep creation and deletion in the same scope. b) Actually run the command provided as arguments; i.e. the actual test :).

```
# Remove the queue on script exit
trap "ipcrm -q $QUEUE" EXIT

# Run the specified command
$@
```

Reply:
P.S. In general I suggest checking the test output when you're not sure what's going on. That should make it clear in this case that the tests aren't actually running.

I find it easiest to use the `ctest` command directly when debugging an individual test, and run only the tests I'm working on. Turning on test output, the command would be something like: `ctest --tests-regex tcp --verbose`

To see the output of the test code running under shadow, you'll need to check the logs in the data directories specified in the test commands. e.g. `blocking-lossy.shadow.data`

Reply:
I forgot to include it, but of course I tried with your full script but tests failed. I will try with the ctest  tomorrow :)

Reply:
With `trap` I obtained the following error
```
3: ** (process:1000): ERROR **: 00:00:01.000: !((msgqueue = msgget(key, 0)) != -1): No such file or directory
3: shadow/src/test/tcp/with_q.sh: line 12:  7419 Trace/breakpoint trap   (core dumped) $@
```

Reply:
I searched again but I keep getting errors. I think the better way is to provide the filename from the env argument, then create and delete the queue in the test with `ftok`. It's not the best but should work ! What do you think ?

Reply:
I don't think the `trap` is what's causing the error. It's just reporting it. Assuming you also added the `$@` at the end, the test is now being run (and failing).

I got the tests working in my local copy of this branch but can't seem to push the commits. Did you check the ['allow edits from maintainers'](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) box? It's also my first time trying to use that feature, so I might be doing something wrong.

Anyway, here's my patch that gets the tests working. Feel free to apply these changes locally, or we can try to figure out the 'allow edits from maintainers' thing.

```
diff --git a/src/test/tcp/test_tcp.c b/src/test/tcp/test_tcp.c
index 2b7a6d1f..ab675a1a 100644
--- a/src/test/tcp/test_tcp.c
+++ b/src/test/tcp/test_tcp.c
@@ -62,32 +62,12 @@ static void _mylog(const char* fileName, const int lineNum, const char* funcName
     fflush(stdout);
 }
 
-static void remove_queue() {
-    assert_true_errno(msgctl(msgqueue, IPC_RMID, NULL) != -1);
-}
-
-// create a System V queue
-// the key can be provide directly as number or put in env variable
-// if the program is running as client, we should remove the message queue at the end
-static void create_msgqueue(const char *queueid) {
-    char *endptr;
-    long int key;
-
-    if (queueid[0] == '$'){
-        queueid = getenv(queueid + 1);
-    }
-
-    errno = 0;
-    key = strtol(queueid, &endptr, 10);
-    // Check for various possible errors
-    if ((errno == ERANGE && (key == LONG_MAX || key == LONG_MIN))
-        || (errno != 0 && key == 0)
-        || endptr == queueid) {
-        perror("strtol");
-        exit(EXIT_FAILURE);
-    }
-
-    assert_true_errno((msgqueue = msgget(key, 0)) != -1);
+// Get the message queue id.
+static void get_msgqueue() {
+    char* id_string = getenv("QUEUE");
+    // FIXME: Validate that integer conversion worked.
+    // FIXME: Should return the queue id. We don't need to use a global anymore.
+    msgqueue = atoi(id_string);
 }
 
 static void queue_send_u16(uint16_t i) {
@@ -824,6 +804,7 @@ int main(int argc, char *argv[]) {
     }
 
     const char *io_mode = argv[1];
+    // FIXME: We don't need this argument anymore.
     const char *queuename = argv[2];
     const char *execution_mode = argv[3];
     iowait_func wait = NULL;
@@ -851,11 +832,11 @@ int main(int argc, char *argv[]) {
             MYLOG("error, client mode also needs a server ip address; see usage");
             return -1;
         }
-        create_msgqueue(queuename);
+        get_msgqueue();
         MYLOG("running client in mode %s", io_mode);
         result = _run_client(wait, argv[4], use_iov);
     } else if(strncasecmp(execution_mode, "server", 6) == 0) {
-        create_msgqueue(queuename);
+        get_msgqueue();
         MYLOG("running server in mode %s", io_mode);
         result = _run_server(wait, use_iov);
     } else {
diff --git a/src/test/tcp/with_q.sh b/src/test/tcp/with_q.sh
index c5b67126..73dbd59b 100755
--- a/src/test/tcp/with_q.sh
+++ b/src/test/tcp/with_q.sh
@@ -5,3 +5,9 @@ set -euo pipefail
 
 # Create Q, and export into environment of child processes
 export QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'`
+
+# Ensure Q is removed on exit
+trap "ipcrm -q $QUEUE" EXIT
+
+# Interpret remaining args as a command to run:
+$@

```

Reply:
Ah ! Ok I understand now, my usage of the IPC interface were really bad ! Thank you @sporksmith 
I allowed maintainers to edit my PR, but I don't know why you can't. Maybe because the PR comes from a fork ?

Reply:
Ok; it might just be that I don't have sufficient permission in the shadow repo to be considered a 'maintainer' for this purpose.

Attaching patches created with `git format-patch` for convenience. Feel free to apply them with `git am` if you like. If it's easier to just manually make the changes again yourself manually that's fine too.

0001-with_q.sh-Run-the-provided-command-and-destroy-q-on-.patch
```
From 3bf59bbe90f93c04fe8846c4a2a7829f171150f7 Mon Sep 17 00:00:00 2001
From: Jim Newsome <jnewsome@torproject.org>
Date: Fri, 15 May 2020 09:24:56 -0500
Subject: [PATCH 1/2] with_q.sh: Run the provided command, and destroy q on
 exit

---
 src/test/tcp/with_q.sh | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/test/tcp/with_q.sh b/src/test/tcp/with_q.sh
index c5b67126..73dbd59b 100755
--- a/src/test/tcp/with_q.sh
+++ b/src/test/tcp/with_q.sh
@@ -5,3 +5,9 @@ set -euo pipefail
 
 # Create Q, and export into environment of child processes
 export QUEUE=`ipcmk --queue | sed 's/[^0-9]*//g'`
+
+# Ensure Q is removed on exit
+trap "ipcrm -q $QUEUE" EXIT
+
+# Interpret remaining args as a command to run:
+$@
-- 
2.26.2
```

0002-test_tcp-Use-queue-ID-from-environment-variable.patch
```From bae2ac7a7489f1aebbc473789069176a029bfe94 Mon Sep 17 00:00:00 2001
From: Jim Newsome <jnewsome@torproject.org>
Date: Fri, 15 May 2020 09:25:19 -0500
Subject: [PATCH 2/2] test_tcp: Use queue ID from environment variable

---
 src/test/tcp/test_tcp.c | 37 +++++++++----------------------------
 1 file changed, 9 insertions(+), 28 deletions(-)

diff --git a/src/test/tcp/test_tcp.c b/src/test/tcp/test_tcp.c
index 2b7a6d1f..ab675a1a 100644
--- a/src/test/tcp/test_tcp.c
+++ b/src/test/tcp/test_tcp.c
@@ -62,32 +62,12 @@ static void _mylog(const char* fileName, const int lineNum, const char* funcName
     fflush(stdout);
 }
 
-static void remove_queue() {
-    assert_true_errno(msgctl(msgqueue, IPC_RMID, NULL) != -1);
-}
-
-// create a System V queue
-// the key can be provide directly as number or put in env variable
-// if the program is running as client, we should remove the message queue at the end
-static void create_msgqueue(const char *queueid) {
-    char *endptr;
-    long int key;
-
-    if (queueid[0] == '$'){
-        queueid = getenv(queueid + 1);
-    }
-
-    errno = 0;
-    key = strtol(queueid, &endptr, 10);
-    // Check for various possible errors
-    if ((errno == ERANGE && (key == LONG_MAX || key == LONG_MIN))
-        || (errno != 0 && key == 0)
-        || endptr == queueid) {
-        perror("strtol");
-        exit(EXIT_FAILURE);
-    }
-
-    assert_true_errno((msgqueue = msgget(key, 0)) != -1);
+// Get the message queue id.
+static void get_msgqueue() {
+    char* id_string = getenv("QUEUE");
+    // FIXME: Validate that integer conversion worked.
+    // FIXME: Should return the queue id. We don't need to use a global anymore.
+    msgqueue = atoi(id_string);
 }
 
 static void queue_send_u16(uint16_t i) {
@@ -824,6 +804,7 @@ int main(int argc, char *argv[]) {
     }
 
     const char *io_mode = argv[1];
+    // FIXME: We don't need this argument anymore.
     const char *queuename = argv[2];
     const char *execution_mode = argv[3];
     iowait_func wait = NULL;
@@ -851,11 +832,11 @@ int main(int argc, char *argv[]) {
             MYLOG("error, client mode also needs a server ip address; see usage");
             return -1;
         }
-        create_msgqueue(queuename);
+        get_msgqueue();
         MYLOG("running client in mode %s", io_mode);
         result = _run_client(wait, argv[4], use_iov);
     } else if(strncasecmp(execution_mode, "server", 6) == 0) {
-        create_msgqueue(queuename);
+        get_msgqueue();
         MYLOG("running server in mode %s", io_mode);
         result = _run_server(wait, use_iov);
     } else {
-- 
2.26.2
```

Reply:
I just see your update, sorry. I think I can change the commit owner and force push to let you as owner of your modifications if you want